sockets/unix_socket.go

```diff
@@ -67,9 +67,13 @@ func WithChown(uid, gid int) SockOption {
 	}
 }
 
-// WithChmod modifies socket file's access mode
+// WithChmod modifies socket file's access mode. The umask is temporarily changed
+// to make sure that the given permissions are applied.
 func WithChmod(mask os.FileMode) SockOption {
 	return func(path string) error {
+		umask := syscall.Umask(0777)
+		defer syscall.Umask(umask)
```

Comment:
So, reading this again, I'm actually somewhat confused now; umask `0777` would mean "ignore all permissions"? Should this be `0000` (i.e.; don't mask anything?); is this actually needed, given that we first _create_ the socket (`umask` applies to _creating_ files), but after that we _chmod_ (to which `umask` doesn't apply)?


```console
$ docker run -it --rm alpine
/ # umask
0022
/ # umask 0777
/ # touch /mysocket && ls -la /mysocket
----------    1 root     root             0 Feb 15 11:21 /mysocket
/ # chmod 0777 /mysocket && ls -la /mysocket
-rwxrwxrwx    1 root     root             0 Feb 15 11:21 /mysocket
```

Reply:
Digging a bit further again; https://github.com/moby/moby/commit/dacae746b70f50dd1f3ea9d40834386b96b6c200 refactored this code to use a `defer`, but the original `umask` change was in https://github.com/moby/moby/pull/5947, which only mentions "fix race condition during socket creation", and after asking for more details;

> It's just a possible security race condition, no actual error.

After scratching my head a bit, I think the intent was to;

1. set the umask to `0777` so that socket is with `0000` permissions
2. chmod to `<desired mode>`

Without the `umask`, the race condition is between step 1. and 2.; step `1.` would create the socket with default permissions (and depend on umask), which could be "less restrictive" than "desired" (step 2.: the actual permissions)

As an alternative, the socket would have to be _created_ with explicit `0000` permissions, but `net.Listen()` does not accept file permissions when creating (I know recall Simon mentioning that)